### PR TITLE
MTLS: Add CA sign method

### PR DESCRIPTION
**Pull request by eloycoto**
This is a commit that implements the Sign method to be used on the
registration API endpoint.

It receives a Certificate Signed Request and sign and return a x509
Certificate signed by the first CA authority.

This PR is part of: ECOPROJECT-401

Signed-off-by: Eloy Coto <eloy.coto@acalustra.com>

### CODE REVIEW

File: internal/mtls/caprovider_secret.go [hunk 1]
```diff
@@ -118,3 +127,55 @@ func (config *CASecretProvider) CreateRegistrationCertificate(name string) (map[
...
+	}
+
+	clientCert := &x509.Certificate{
+		Signature:          CSR.Signature,
+		SignatureAlgorithm: CSR.SignatureAlgorithm,
+		PublicKeyAlgorithm: CSR.PublicKeyAlgorithm,
+		PublicKey:          CSR.PublicKey,
+		SerialNumber:       big.NewInt(time.Now().Unix()),
+		Subject:            CSR.Subject,
+		NotBefore:          time.Now(),
```
Comment (pkliczewski) [inline internal/mtls/caprovider_secret.go:158]: In the past this was problematic due to time drift on different machines. We do not require ntpd nor chronyd. In the past we decided to set it to `time.Now() - 24h` to mitigate `certificate not valid yet`.

File: internal/mtls/ca.go [hunk 2]
```diff
@@ -19,6 +20,11 @@ const (
 	regClientSecretNameRandomLen = 10
 	regClientSecretNamePrefix    = "reg-client-ca"
 	regClientSecretLabelKey      = "reg-client-ca"
+
+	YggdrasilRegisterAuth = 1
+	YggdrasilCompleteAuth = 0
+
+	defaultDaysToExpireClientCertificate = 7
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply internal/mtls/ca.go:27]: It's already: 
https://github.com/jakub-dzon/k4e-operator/pull/59/files#diff-5a67b56a16375284623f928642038d3a9234c78a1827d835e85688084be8c1dfR70


https://github.com/jakub-dzon/k4e-operator/pull/59/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R222

File: internal/mtls/ca.go [hunk 3]
```diff
@@ -57,6 +66,26 @@ func NewMTLSConfig(client client.Client, namespace string, domains []string, loc
...
+func (conf *TLSConfig) SetClientExpiration(days int) error {
+	if days <= 0 {
+		return fmt.Errorf("Cannot set 1 day expiration time")
+	}
+	conf.clientExpirationDays = days
+	return nil
+}
+
+// SignCSR sign the given CSRPem using the first CA provider in use.
+func (conf *TLSConfig) SignCSR(CSRPem string, commonName string) ([]byte, error) {
```
Comment (masayag) [inline internal/mtls/ca.go:79]: the next PR will make use of this function as part of the registration call ?

**Review by masayag — COMMENTED** [review]
Overall, this change looks good to me.
We should probably start thinking on supporting namespaces other than the default for this change to be more effective.

File: internal/mtls/ca.go [hunk 4]
```diff
@@ -19,6 +20,11 @@ const (
 	regClientSecretNameRandomLen = 10
 	regClientSecretNamePrefix    = "reg-client-ca"
 	regClientSecretLabelKey      = "reg-client-ca"
+
+	YggdrasilRegisterAuth = 1
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply internal/mtls/ca.go:24]: The  yggdrasil will be deleted in the next PR. Issues with cycle import on the next PR.

File: main.go [hunk 5]
```diff
@@ -216,6 +219,11 @@ func main() {
 		MTLSconfig := mtls.NewMTLSConfig(mgr.GetClient(), operatorNamespace,
```
Comment (jakub-dzon) [inline main.go:219]: I haven't spotted that before, but it `MTLSconfig` should start with a lower-case letter.

File: internal/mtls/caprovider_secret.go [hunk 6]
```diff
@@ -118,3 +127,55 @@ func (config *CASecretProvider) CreateRegistrationCertificate(name string) (map[
 	}
 	return res, nil
 }
+
+// SignCSR sign a new CertificateRequest and returns the PEM certificate.
+// This function is going to be used a lot, so using config.latestCA ensure
+// that APIServer is not overloaded with that.
```
Comment (jakub-dzon) [inline internal/mtls/caprovider_secret.go:133]: Wouldn't k8s client caching take care of that?

Comment (eloycoto) [reply internal/mtls/caprovider_secret.go:133]: No, I tested and it's not caching this at all, and I do not think that it should TBH.

**Review by jakub-dzon — APPROVED** [review]

**Pull request merged**